### PR TITLE
Add `Node` and `Graph` classes to allow organisation of data collection and database insertion

### PR DESCRIPTION
**Pull request by d-j-hatton**
There are already `ProcessNode` and `ProcessGraph` classes to mimic the structure of a Relion project and help keep track of which jobs have run/been scheduled. Propose that `DBNode` and `DBGraph` be added in order to get data collected from the `ProcessNode`s to the correct tables in the database model introduced in #111. Refactor to have both types of node inherit from a base class to allow data transfer between the two. The project should then be able to be described as a graph structure, one part of which follow the Relion job structure, which is linked to another part that mimics the structure of the database schema.

### CODE REVIEW

File: src/relion/node/graph.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+            self._call_returns[node.nodeid] = node()
+            self._called_nodes.append(node.nodeid)
+
+        for next_node in node:
+            next_node.environment.update_prop(node.environment.propagate)
+            try:
+                next_traffic = node._link_traffic[next_node.nodeid]
+            except KeyError:
+                next_traffic = {}
```
Comment (Anthchirp) [inline src/relion/node/graph.py:181]:
```suggestion
            next_traffic = node._link_traffic.get(next_node.nodeid, {})
```

File: tests/dbmodel/test_dbnode.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    return table
+
+
+@pytest.fixture
+def ctf_table(proj):
+    table = modeltables.CTFTable()
+    ctf_res = proj.ctffind["job003"]
+    ctf_db_entries = proj.ctffind.db_unpack(ctf_res)
+    for i, entry in enumerate(ctf_db_entries):
+        table.add_row({**entry, **{"motion_correction_id": i + 1}})
```
Comment (Anthchirp) [inline tests/dbmodel/test_dbnode.py:29]: Is this equivalent to 
```python
table.add_row({**entry, "motion_correction_id": i + 1})
```

File: docs/dbmodel.rst [hunk 3]
```diff
@@ -49,3 +49,32 @@ In addition, a column can be set to auto-increment with the ``counters`` key wor
 This is used, for example, in the ``MotionCorrection`` table which contains an ``image_number`` column which 
 needs to be a unique integer for each micrograph. An appendable column may be specified with the ``append`` 
 keyword. For these columns values are appended to a list rather than updated.
+
+------
+DBNode
+------
+
+A ``DBNode`` is a ``Node`` implemented with a specific call. Calling a ``DBNode`` will perform an insert into 
+a ``Table`` associated with the node after doing a few checks. If ``"check_for`` is present in the node's
```
Comment (Anthchirp) [inline docs/dbmodel.rst:58]:
```suggestion
a ``Table`` associated with the node after doing a few checks. If ``check_for`` is present in the node's 
```
would like to note that the entire section is very dense :slightly_smiling_face: 

With the further description in the other file I believe this is sufficiently complex to benefit from docstrings and auto-generated documentation.

File: src/relion/_parser/class2D.py [hunk 4]
```diff
@@ -185,3 +187,22 @@ def top_twenty(self, dictionary):
             temp_list.reverse()
             return_dict[item] = temp_list
         return return_dict
+
+    @staticmethod
+    def db_unpack(particle_class):
+        res = []
```
Comment (Anthchirp) [inline src/relion/_parser/class2D.py:193]: in some cases you do individual appends, in others you have list comprehensions

File: src/relion/_parser/cryolo.py [hunk 5]
```diff
@@ -69,12 +77,49 @@ def _load_job_directory(self, jobdir):
...
+                        (
+                            x
+                            for x in micrograph_names
+                            if str(mic.relative_to(self._basepath / jobdir)).replace(
+                                "_autopick.star", ".mrc"
+                            )
+                            in x
+                        ),
+                        None,
+                    ),
```
Comment (Anthchirp) [inline src/relion/_parser/cryolo.py:100]: Could this be extracted into a function?
(also: a `print` above)

**Review by Anthchirp — APPROVED** [review]
Clearly a lot of effort has gone into this. I would feel more comfortable if we knew more about where relion is going with its python interface and we had a plan on how we are going to merge these two in the future.

For now though this is a welcome solution to an existing problem :+1:

File: docs/dbmodel.rst (hunk 3 above)
Comment (d-j-hatton) [reply docs/dbmodel.rst]: I agree the documentation needs improving. I'll try and make some improvements in the next PR and then do some more once we have a better idea of how things will change once the Relion python interface is available

**Pull request merged**